src/Agent.Listener/SelfUpdater.cs

```diff
@@ -141,95 +141,150 @@ private async Task DownloadLatestAgent(CancellationToken token)
+                    if (_targetPackage.Platform.StartsWith("win"))
+                    {
+                        archiveFile = Path.Combine(latestAgentDirectory, $"agent{agentSuffix}.zip");
+                    }
+                    else
+                    {
+                        archiveFile = Path.Combine(latestAgentDirectory, $"agent{agentSuffix}.tar.gz");
+                    }
+
+                    try
                     {
-                        using (Stream result = await httpClient.GetStreamAsync(_targetPackage.DownloadUrl))
+                        // delete existing .zip file
+                        if (File.Exists(archiveFile))
                         {
-                            //81920 is the default used by System.IO.Stream.CopyTo and is under the large object heap threshold (85k). 
-                            await result.CopyToAsync(fs, 81920, token);
-                            await fs.FlushAsync(token);
+                            Trace.Verbose("Pre-download: deleting latest agent package zip '{0}'", archiveFile);
+                            IOUtil.DeleteFile(archiveFile);
                         }
+
+                        break;
+                    }
+                    catch (Exception ex)
+                    {
+                        // couldn't delete the file for whatever reason, so generate another name
+                        Trace.Warning("Pre-download: failed to delete agent package zip '{0}'. Exception: {1}", archiveFile, ex);
+                        agentSuffix++;
                     }
                 }
 
-                if (archiveFile.EndsWith(".zip", StringComparison.OrdinalIgnoreCase))
-                {
-                    ZipFile.ExtractToDirectory(archiveFile, latestAgentDirectory);
-                }
-                else if (archiveFile.EndsWith(".tar.gz", StringComparison.OrdinalIgnoreCase))
+                Trace.Info($"Attempt {attempt}: save latest agent into {archiveFile}.");
+
+                bool gettingPackageSucceeded = false;
+
+                try
                 {
-                    string tar = WhichUtil.Which("tar", trace: Trace);
-                    if (string.IsNullOrEmpty(tar))
-                    {
-                        throw new NotSupportedException($"tar -xzf");
-                    }
+                    bool downloadSucceeded = false;
 
-                    // tar -xzf
-                    using (var processInvoker = HostContext.CreateService<IProcessInvoker>())
+                    using (var httpClient = new HttpClient(HostContext.CreateHttpClientHandler()))
                     {
-                        processInvoker.OutputDataReceived += new EventHandler<ProcessDataReceivedEventArgs>((sender, args) =>
+                        // Set a 15-minute timeout per agent package download attempt
+                        httpClient.Timeout = TimeSpan.FromMinutes(15);
```

Comment:
I don't think you want to set this, the client.timeout is the timeout for a client received header from server not the entire downloading time.

Reply:
You're right. The MSDN documentation was kind of vague of what the httpClient.Timeout was used for. Fixed.